### PR TITLE
[ENHANCEMENT] add z-loss (improved version)

### PR DESCRIPTION
**Pull request by wdevazelhes**
z-loss regularization was shown to stabilize training by preventing logits in the last layer to explode: https://arxiv.org/abs/2309.14322

This PR:

- allows one to add zloss regularization
- logs the z-loss

Note: this PR takes over #1270 by logging the z-loss more cleanly leveraging the already existing of auxiliary losses present for MoE

### CODE REVIEW

**Comment by Phlip79** [comment]
**We are changing our review process and marking all open, unlabeled PRs as draft. This change will go in effect starting once https://github.com/NVIDIA/Megatron-LM/pull/3659 is merged.**

Moving forward, all PRs will be required to start as draft PRs. If you wish to get your PR merged, mark your PR as “Ready for review”. Read more about the new process at [submit.md](https://github.com/Phlip79/Megatron-LM/blob/d3763629bcdd04acca52477af7eb8795090426ff/docs/developer/submit.md).

**Comment by yaox12** [comment]
Close this PR since I believe z-loss has been added to Megatron Core MoE.
